File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-java/src/test/java/com/opspilot/ResilientChatModelTest.java
package com.opspilot;

import com.opspilot.domain.Exceptions;
import com.opspilot.llm.ChatMessage;
import com.opspilot.llm.ChatModel;
import com.opspilot.llm.ModelResponse;
import com.opspilot.llm.ResilientChatModel;
import com.opspilot.tools.Tool;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class ResilientChatModelTest {

    private static ChatModel alwaysFails(AtomicInteger counter) {
        return (messages, tools) -> {
            counter.incrementAndGet();
            throw new RuntimeException("provider indisponivel");
        };
    }

    private static ChatModel alwaysSucceeds(String answer) {
        return (messages, tools) -> ModelResponse.text(answer);
    }

    @Test
    void succeedsOnPrimaryWithoutTouchingBackup() {
        AtomicInteger backupCalls = new AtomicInteger();
        ChatModel primary = alwaysSucceeds("ok");
        ChatModel backup = (messages, tools) -> {
            backupCalls.incrementAndGet();
            return ModelResponse.text("nao deveria chegar aqui");
        };

        ResilientChatModel resilient = new ResilientChatModel(primary, backup);
        ModelResponse response = resilient.invoke(List.of(ChatMessage.user("oi")), List.of());

        assertThat(response.content()).isEqualTo("ok");
        assertThat(backupCalls.get()).isZero();
    }

    @Test
    void fallsBackToBackupWhenPrimaryExhaustsRetries() {
        AtomicInteger primaryCalls = new AtomicInteger();
        ChatModel primary = alwaysFails(primaryCalls);
        ChatModel backup = alwaysSucceeds("resposta do backup");

        ResilientChatModel resilient = new ResilientChatModel(primary, backup);
        ModelResponse response = resilient.invoke(List.of(ChatMessage.user("oi")), List.of());

        assertThat(response.content()).isEqualTo("resposta do backup");
        assertThat(primaryCalls.get()).isEqualTo(ResilientChatModel.RETRY_ATTEMPTS);
    }

    @Test
    void throwsModelUnavailableWhenNoBackupConfigured() {
        ResilientChatModel resilient = new ResilientChatModel(alwaysFails(new AtomicInteger()), null);

        assertThatThrownBy(() -> resilient.invoke(List.of(ChatMessage.user("oi")), List.<Tool>of()))
                .isInstanceOf(Exceptions.ModelUnavailableException.class);
    }

    @Test
    void throwsModelUnavailableWhenBothPrimaryAndBackupFail() {
        ResilientChatModel resilient = new ResilientChatModel(
                alwaysFails(new AtomicInteger()), alwaysFails(new AtomicInteger()));

        assertThatThrownBy(() -> resilient.invoke(List.of(ChatMessage.user("oi")), List.<Tool>of()))
                .isInstanceOf(Exceptions.ModelUnavailableException.class);
    }
}

File: modulo04-criacao-de-agentes-autonomos-novo/opspilot-go/llm/resilient_chat_model_test.go
package llm

import (
	"errors"
	"sync/atomic"
	"testing"

	"opspilot/domain"
	"opspilot/tools"
)

func alwaysFails(counter *atomic.Int32) ChatModel {
	return ChatModelFunc(func(messages []ChatMessage, availableTools []tools.Tool) (ModelResponse, error) {
		counter.Add(1)
		return ModelResponse{}, errors.New("provider indisponivel")
	})
}

func alwaysSucceeds(answer string) ChatModel {
	return ChatModelFunc(func(messages []ChatMessage, availableTools []tools.Tool) (ModelResponse, error) {
		return TextResponse(answer), nil
	})
}

func TestSucceedsOnPrimaryWithoutTouchingBackup(t *testing.T) {
	var backupCalls atomic.Int32
	primary := alwaysSucceeds("ok")
	backup := ChatModelFunc(func(messages []ChatMessage, availableTools []tools.Tool) (ModelResponse, error) {
		backupCalls.Add(1)
		return TextResponse("nao deveria chegar aqui"), nil
	})

	resilient := NewResilientChatModel(primary, backup)
	response, err := resilient.Invoke([]ChatMessage{UserMessage("oi")}, nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if response.Content != "ok" {
		t.Fatalf("expected 'ok', got %q", response.Content)
	}
	if backupCalls.Load() != 0 {
		t.Fatalf("expected backup untouched, got %d calls", backupCalls.Load())
	}
}

func TestFallsBackToBackupWhenPrimaryExhaustsRetries(t *testing.T) {
	var primaryCalls atomic.Int32
	primary := alwaysFails(&primaryCalls)
	backup := alwaysSucceeds("resposta do backup")

	resilient := NewResilientChatModel(primary, backup)
	response, err := resilient.Invoke([]ChatMessage{UserMessage("oi")}, nil)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if response.Content != "resposta do backup" {
		t.Fatalf("expected backup response, got %q", response.Content)
	}
	if int(primaryCalls.Load()) != RetryAttempts {
		t.Fatalf("expected %d primary calls, got %d", RetryAttempts, primaryCalls.Load())
	}
}

func TestThrowsModelUnavailableWhenNoBackupConfigured(t *testing.T) {
	var counter atomic.Int32
	resilient := NewResilientChatModel(alwaysFails(&counter), nil)

	_, err := resilient.Invoke([]ChatMessage{UserMessage("oi")}, nil)
	var unavailable *domain.ModelUnavailableError
	if !errors.As(err, &unavailable) {
		t.Fatalf("expected ModelUnavailableError, got %T (%v)", err, err)
	}
}

func TestThrowsModelUnavailableWhenBothPrimaryAndBackupFail(t *testing.T) {
	var primaryCounter, backupCounter atomic.Int32
	resilient := NewResilientChatModel(alwaysFails(&primaryCounter), alwaysFails(&backupCounter))

	_, err := resilient.Invoke([]ChatMessage{UserMessage("oi")}, nil)
	var unavailable *domain.ModelUnavailableError
	if !errors.As(err, &unavailable) {
		t.Fatalf("expected ModelUnavailableError, got %T (%v)", err, err)
	}
}
